File: 算法/代码随想录算法PDF/00调试源代码/131.分割回文串.cpp
/*
 * @lc app=leetcode.cn id=131 lang=cpp
 *
 * [131] 分割回文串
 */

// @lc code=start
class Solution {
public:
    vector<vector<string>> res;
    vector<string> path;
    bool isPalindrome(string& s, int start, int end) {
        while (start <= end) {
            if (s[start] != s[end]) return false;
            start++;
            end--;
        }
        return true;
    }
    void backtracking(string& s, int startIndex) {
        if (startIndex >= s.size()) {
            res.push_back(path);
            return ;
        }
        for (int i = startIndex; i < s.size(); i++) {
            if (isPalindrome(s, startIndex, i)) {
                path.push_back(s.substr(startIndex, i - startIndex + 1));
                backtracking(s, i + 1);
                path.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        backtracking(s, 0);
        return res;
    }
};
// @lc code=end

File: 算法/代码随想录算法PDF/00调试源代码/131.分割回文串.py
#
# @lc app=leetcode.cn id=131 lang=python3
#
# [131] 分割回文串
#

# @lc code=start
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        res = []
        path = []
        def isPalindrome(s, start, end):
            while start <= end:
                if s[start] != s[end]: return False
                start += 1
                end -= 1
            return True

        def backtracking(s, startIndex):
            nonlocal res, path
            if startIndex >= len(s):
                res.append(path[:])
                return
            for i in range(startIndex, len(s)):
                if isPalindrome(s, startIndex, i):
                    path.append(s[startIndex : i + 1])
                    backtracking(s, i + 1)
                    path.pop()
        
        backtracking(s, 0)
        return res
# @lc code=end

